### PR TITLE
adding tests and Reordering WCS ordering

### PR DESCRIPTION
**Pull request by abit2**
I have to write tests for 4d cube.
Tests for 3d cube covered

### CODE REVIEW

File: sunpycube/cube/tests/test_cube.py [hunk 1]
```diff
@@ -107,137 +98,154 @@ def test_choose_x_slice():
...
+
 def test_wavelength_axis():
     cube_wavelength_axis = cube.wavelength_axis()
     cubem_wavelength_axis = cubem.wavelength_axis()
     f1, u1 = cube_wavelength_axis.value, cube_wavelength_axis.unit
     f2, u2 = cubem_wavelength_axis.value, cubem_wavelength_axis.unit
     # the e-11 are the conversions from angstrom to meters
-    assert np.allclose(f1, [0, 2.0e-11, 4.0e-11])
-    assert np.allclose(f2, [0, 2.0e-11, 4.0e-11])
+    assert np.allclose(f1, [0.00000000e+00,  3.00000000e-11,  6.00000000e-11])
```
Comment (ehsteve) [inline sunpycube/cube/tests/test_cube.py:137]: is there an assert_all_close function in py.test

File: sunpycube/cube/tests/test_cube.py [hunk 2]
```diff
@@ -107,137 +98,154 @@ def test_choose_x_slice():
...
               cube[:, 0:0.4 * u.Angstrom:0.4], cube[1, 0.2 * u.Angstrom],
-              cube[0.5 * u.min, 0:2], cube[0.5 * u.min, 0:0.4 * u.Angstrom:0.4]
+              cube[0.5 * u.deg, 0:2], cube[0.5 * u.deg, 0:0.4 * u.Angstrom:0.4]
               ]
 
-    types = [np.ndarray, c.Cube, c.Cube, np.ndarray, GenericMap, GenericMap,
-             LightCurve, c.Cube, c.Cube, np.ndarray, Spectrum, Spectrum]
+    types = [np.ndarray, Cube, Cube, np.ndarray, GenericMap, GenericMap,
+             LightCurve, Cube, Cube, np.ndarray, Spectrum, Spectrum]
     for (s, t) in zip(slices, types):
```
Comment (ehsteve) [inline sunpycube/cube/tests/test_cube.py]: you should use py.test parameterize to do this.

File: sunpycube/cube/cube_utils.py [hunk 3]
```diff
@@ -517,8 +517,19 @@ def get_cube_from_sequence(cubesequence, item):
         The item to get from the cube
     """
     if isinstance(item, int):
-        return cubesequence.data[item]
-    return cubesequence.data[item[0]][item[1::]]
+        result = cubesequence.data[item]
+
+    if isinstance(item, slice):
+        data = cubesequence.data[item]
+        print(type(data))
```
Comment (DanRyanIrish) [inline sunpycube/cube/cube_utils.py]: This print statement should be removed.

File: sunpycube/cube/cube_utils.py [hunk 4]
```diff
@@ -267,19 +267,19 @@ def getitem_3d(cube, item):
         The item to get from the cube
     """
     axes = cube.axes_wcs.wcs.ctype
-    slice_to_map = (axes[-2] != 'WAVE' and
+    slice_to_map = (axes[1] != 'WAVE' and
```
Comment (DanRyanIrish) [inline sunpycube/cube/cube_utils.py:270]: For all these magic indexing numbers, can you add comments explaining why you are using them?

**Comment by DanRyanIrish** [comment]
As a general comment, it would be good to have some more comments to make it easier to follow you the is (or supposed to be) doing.

Will take a closer look as this PR soon.

File: sunpycube/cube/cube_utils.py (hunk 4 above)
Comment (abit2) [reply sunpycube/cube/cube_utils.py:270]: the change in orientation.
before [space, space, wave, time] 
after    [time, wave, space, space]

File: sunpycube/cube/cube_utils.py [hunk 5]
```diff
@@ -267,19 +267,19 @@ def getitem_3d(cube, item):
...
-                          and axes[-2] == 'WAVE')
-                         or (axes[-1] == 'WAVE' and
+                          and axes[1] == 'WAVE')
+                         or (axes[0] == 'WAVE' and
                              iter_isinstance(item, (slice, int, int))))
     slice_to_spectrogram = (iter_isinstance(item, (slice, slice, int)) and
-                            axes[-2] == 'WAVE')
-    slice_to_lightcurve = (axes[-2] == 'WAVE' and
+                            axes[1] == 'WAVE')
+    slice_to_lightcurve = (axes[1] == 'WAVE' and
```
Comment (abit2) [inline sunpycube/cube/cube_utils.py:282]: same as above comment in these changes

File: sunpycube/cube/cube_utils.py [hunk 6]
```diff
@@ -393,6 +393,8 @@ def pixelize_slice(item, wcs, _source='cube'):
         result = tuple(result)
     elif isinstance(item, u.Quantity):
         result = convert_point(item.value, item.unit, wcs, 0)
+    elif isinstance(item, slice):
+        result = _convert_slice(item, wcs, 0, _source=_source)
```
Comment (abit2) [inline sunpycube/cube/cube_utils.py:397]: case of slice was not handled so added that.

File: sunpycube/cube/cube_utils.py [hunk 7]
```diff
@@ -425,19 +427,17 @@ def convert_point(value, unit, wcs, axis, _source='cube'):
         value = value.value
         unit = value.unit
     if _source == 'cube':
-        wcsaxis = -1 - axis if wcs.oriented or not wcs.was_augmented \
+        wcsaxis = -1 - axis if wcs.oriented and not wcs.was_augmented \
```
Comment (abit2) [inline sunpycube/cube/cube_utils.py:430]: it should be and as wcs.was_augmented = false means, wcs object has dimension 3 so wcsaxis = -1-axis not -2-axis and similary for all cases.

File: sunpycube/cube/cube_utils.py [hunk 8]
```diff
@@ -467,7 +467,7 @@ def _convert_slice(item, wcs, axis, _source='cube'):
         opposite WCS convention)
     """
     if _source == 'cube':
-        wcs_ax = -2 - axis if wcs.was_augmented and not wcs.oriented \
+        wcs_ax = -2 - axis if wcs.was_augmented or not wcs.oriented \
```
Comment (abit2) [inline sunpycube/cube/cube_utils.py:470]: even if wcs.oriented is true /false if wcs.was_augmented is true than the wcs object has 4 dim so wcs_ax=-2-axis should be called

File: sunpycube/cube/cube_utils.py (hunk 3 above)
Comment (abit2) [reply sunpycube/cube/cube_utils.py]: handling slice item type

File: sunpycube/cube/cube_utils.py [hunk 9]
```diff
@@ -517,8 +517,19 @@ def get_cube_from_sequence(cubesequence, item):
...
+    if isinstance(item, slice):
+        data = cubesequence.data[item]
+        print(type(data))
+        meta = cubesequence.meta
+        result = cubesequence._new_instance(data, meta)
+
+    if isinstance(item, tuple):
+        data = cubesequence.data[item[0]][item[1::]]
+        meta = cubesequence.meta
+        result = cubesequence._new_instance(data, meta)
```
Comment (abit2) [inline sunpycube/cube/cube_utils.py:531]: returning cubesequence object now earlier returned cube object

File: sunpycube/cube/datacube.py [hunk 10]
```diff
@@ -193,7 +193,7 @@ def _choose_wavelength_slice(self, offset):
         if self.data.ndim == 4:
             raise cu.CubeError(4, "Can only work with 3D cubes")
 
-        axis = 1 if self.axes_wcs.wcs.ctype[-1] in ['TIME', 'UTC'] else 0
+        axis = -2 if self.axes_wcs.wcs.ctype[0] in ['TIME', 'UTC'] else -1
```
Comment (abit2) [inline sunpycube/cube/datacube.py:197]: ctype[0] will be time so changing from 1 to -2 as orientation is changed. similarly for -1.

File: sunpycube/cube/datacube.py [hunk 11]
```diff
@@ -221,14 +221,14 @@ def _choose_x_slice(self, offset):
             wavelength to the one specified.
         """
         arr = None
-        axis = 1 if self.axes_wcs.wcs.ctype[-2] != 'WAVE' else 2
+        axis = 0
```
Comment (abit2) [inline sunpycube/cube/datacube.py:225]: x axis will always be 0 th index of data.

File: sunpycube/cube/datacube.py [hunk 12]
```diff
@@ -221,14 +221,14 @@ def _choose_x_slice(self, offset):
...
         length = self.data.shape[axis]
         if isinstance(offset, int) and offset >= 0 and offset < length:
             arr = self.data.take(offset, axis=axis)
 
         if isinstance(offset, u.Quantity):
-            unit = self.axes_wcs.wcs.cunit[-1 - axis]
-            delta = self.axes_wcs.wcs.cdelt[-1 - axis] * unit
+            unit = self.axes_wcs.wcs.cunit[-1]
```
Comment (abit2) [inline sunpycube/cube/datacube.py:231]: -1-0 so changed it to -1

File: sunpycube/cube/datacube.py [hunk 13]
```diff
@@ -360,15 +360,15 @@ def slice_to_spectrum(self, *coords, **kwargs):
...
             if pixels[i] is None:
                 if i == 0:
-                    sumaxis = 1 if axis == 0 else 0
+                    sumaxis = 1 if axis == -1 else 0
                 else:
                     sumaxis = 1 if i == 2 else i
                 data = data.sum(axis=sumaxis)
-
+                mask = mask.sum(axis=sumaxis)
+        kwargs.update({'uncertainty': errors, 'mask': mask})
```
Comment (abit2) [inline sunpycube/cube/datacube.py:372]: dimension of mask and data were not matching so hacked my way through this ;)

File: sunpycube/cube/datacube.py [hunk 14]
```diff
@@ -507,14 +507,14 @@ def wavelength_axis(self):
...
+        delta = self.axes_wcs.wcs.cdelt[axis]
+        crpix = self.axes_wcs.wcs.crpix[axis]
+        crval = self.axes_wcs.wcs.crval[axis]
         start = crval - crpix * delta
-        stop = start + self.data.shape[axis] * delta
-        cunit = u.Unit(self.axes_wcs.wcs.cunit[-1 - axis])
-        return np.linspace(start, stop, num=self.data.shape[axis]) * cunit
+        stop = start + self.data.shape[-1 - axis] * delta
+        cunit = u.Unit(self.axes_wcs.wcs.cunit[axis])
+        return np.linspace(start, stop, num=self.data.shape[-1 - axis]) * cunit
```
Comment (abit2) [inline sunpycube/cube/datacube.py:518]: the same orientation changes in the above lines

File: sunpycube/cube/datacube.py [hunk 15]
```diff
@@ -549,8 +549,16 @@ def __init__(self, data_list, meta=None):
         if not all(isinstance(data, Cube) for data in data_list):
             raise ValueError("data list should be of cube object")
         self.data = data_list
+        self.len = len(data_list)
         self.meta = meta
 
+    @classmethod
+    def _new_instance(cls, data_list, meta=None):
```
Comment (abit2) [inline sunpycube/cube/datacube.py]: class method to make new cubesequence used in indexing

File: sunpycube/cube/cube_utils.py (hunk 4 above)
Comment (DanRyanIrish) [reply sunpycube/cube/cube_utils.py:270]: Hi @abit2.  I meant comments in the code based itself using '#' :)

File: sunpycube/cube/cube_utils.py (hunk 7 above)
Comment (abit2) [reply sunpycube/cube/cube_utils.py:430]: @DanRyanIrish making this not wcs.was_augmented seems more right to me

Comment (abit2) [reply sunpycube/cube/cube_utils.py:430]: like remove the wcs.oriented part and just keeping 
```python
if not wcs.was_augemented:
    ......
```

File: sunpycube/cube/cube_utils.py (hunk 4 above)
Comment (dpshelio) [reply sunpycube/cube/cube_utils.py:270]: I was thinking the same :)

File: sunpycube/cube/tests/test_cube.py (hunk 1 above)
Comment (Cadair) [reply sunpycube/cube/tests/test_cube.py:137]: there is `assert_allclose` or `asserr_quantity_allclose`

**Comment by DanRyanIrish** [comment]
@ehsteve @Cadair Did either of you get a chance to look quickly at this PR for any final comments before merging?  If you wont have time I suggest we just merge it once the conflicts have been resolved.

File: sunpycube/cube/tests/test_cube.py (hunk 1 above)
Comment (abit2) [reply sunpycube/cube/tests/test_cube.py:137]: @Cadair is asset_allclose in pytest or did you mean numpy.testing.assert_allclose?

**Comment by DanRyanIrish** [comment]
👍

**Pull request merged**